diff -Naur a/gcc/config/mips/mips.md b/gcc/config/mips/mips.md
--- a/gcc/config/mips/mips.md	2019-02-22 19:38:43.000000000 +0300
+++ b/gcc/config/mips/mips.md	2026-06-03 06:30:48.088696208 +0300
@@ -3142,7 +3142,24 @@
 ;;  ...................
 ;;
 
-(define_insn "clz<mode>2"
+(define_expand "clzsi2"
+  [(set (match_operand:SI 0 "register_operand")
+	(clz:SI (match_operand:SI 1 "register_operand")))]
+  "ISA_HAS_CLZ_CLO || TARGET_MIPS5900"
+  {
+    if (TARGET_MIPS5900)
+      {
+	rtx scratch = gen_rtx_SCRATCH (SImode);
+	
+	emit_insn (gen_rtx_PARALLEL (VOIDmode,
+	  gen_rtvec (2,
+	    gen_rtx_SET (operands[0], gen_rtx_CLZ (SImode, operands[1])),
+	    gen_rtx_CLOBBER (VOIDmode, scratch))));
+	DONE;
+      }
+  })
+
+(define_insn "*clz<mode>2"
   [(set (match_operand:GPR 0 "register_operand" "=d")
 	(clz:GPR (match_operand:GPR 1 "register_operand" "d")))]
   "ISA_HAS_CLZ_CLO"
@@ -3150,6 +3167,15 @@
   [(set_attr "type" "clz")
    (set_attr "mode" "<MODE>")])
 
+(define_insn "*clzsi2_r5900"
+  [(set (match_operand:SI 0 "register_operand" "=&d")
+	(clz:SI (match_operand:SI 1 "register_operand" "d")))
+   (clobber (match_scratch:SI 2 "=&d"))]
+  "TARGET_MIPS5900"
+  "srl\t%2,%1,31\;plzcw\t%0,%1\;addiu\t%0,%0,1\;movn\t%0,$zero,%2"
+  [(set_attr "type" "clz")
+   (set_attr "mode" "SI")])
+
 ;;
 ;;  ...................
 ;;
@@ -3158,7 +3184,24 @@
 ;;  ...................
 ;;
 
-(define_insn "ctz<mode>2"
+(define_expand "ctzsi2"
+  [(set (match_operand:SI 0 "register_operand")
+	(ctz:SI (match_operand:SI 1 "register_operand")))]
+  "ISA_HAS_CTZ_CTO || TARGET_MIPS5900"
+  {
+    if (TARGET_MIPS5900)
+      {
+	rtx scratch = gen_rtx_SCRATCH (SImode);
+	
+	emit_insn (gen_rtx_PARALLEL (VOIDmode,
+	  gen_rtvec (2,
+	    gen_rtx_SET (operands[0], gen_rtx_CTZ (SImode, operands[1])),
+	    gen_rtx_CLOBBER (VOIDmode, scratch))));
+	DONE;
+      }
+  })
+
+(define_insn "*ctz<mode>2"
   [(set (match_operand:GPR 0 "register_operand" "=d")
 	(ctz:GPR (match_operand:GPR 1 "register_operand" "d")))]
   "ISA_HAS_CTZ_CTO"
@@ -3166,6 +3209,14 @@
   [(set_attr "type" "clz")
    (set_attr "mode" "<MODE>")])
 
+(define_insn "*ctzsi2_r5900"
+  [(set (match_operand:SI 0 "register_operand" "=&d")
+	(ctz:SI (match_operand:SI 1 "register_operand" "d"))) ;; Вот здесь теперь СТРОГО ctz:SI!
+   (clobber (match_scratch:SI 2 "=&d"))]
+  "TARGET_MIPS5900"
+  "subu\t%2,$zero,%1\;and\t%2,%1,%2\;addiu\t%2,%2,-1\;sltiu\t%0,%1,1\;plzcw\t%2,%2\;sll\t%0,%0,31\;xori\t%2,%2,31\;sra\t%0,%0,31\;or\t%0,%0,%2"
+  [(set_attr "type" "clz")
+   (set_attr "mode" "SI")])
 
 ;;
 ;;  ...................
diff -Naur a/libgcc/config/mips/r5900/_clrsbdi2.S b/libgcc/config/mips/r5900/_clrsbdi2.S
--- a/libgcc/config/mips/r5900/_clrsbdi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_clrsbdi2.S	2026-06-03 18:52:11.732601134 +0300
@@ -0,0 +1,39 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__clrsbdi2
+	.hidden	__clrsbdi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__clrsbdi2
+	.type	__clrsbdi2, @function
+__clrsbdi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	srl     $8, $4, 31
+	srl     $9, $5, 31
+	plzcw   $2, $4
+	plzcw   $3, $5
+	addiu   $2, $2, 1
+	xor     $8, $8, $9
+	movn    $2, $zero, $8
+	addiu   $3, $3, 0
+	slti    $9, $3, 31
+	movn    $2, $zero, $9
+	jr	$31
+	addu    $2, $3, $2
+	.set	macro
+	.set	reorder
+
+	.end	__clrsbdi2
+	.size	__clrsbdi2, .-__clrsbdi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/r5900/_clrsbsi2.S b/libgcc/config/mips/r5900/_clrsbsi2.S
--- a/libgcc/config/mips/r5900/_clrsbsi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_clrsbsi2.S	2026-06-03 18:52:19.739550845 +0300
@@ -0,0 +1,30 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__clrsbsi2
+	.hidden	__clrsbsi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__clrsbsi2
+	.type	__clrsbsi2, @function
+__clrsbsi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	plzcw   $2, $4
+	jr	$31
+	addiu	$2, $2, 0
+	.set	macro
+	.set	reorder
+
+	.end	__clrsbsi2
+	.size	__clrsbsi2, .-__clrsbsi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/r5900/_clzdi2.S b/libgcc/config/mips/r5900/_clzdi2.S
--- a/libgcc/config/mips/r5900/_clzdi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_clzdi2.S	2026-06-03 18:52:28.399496129 +0300
@@ -0,0 +1,37 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__clzdi2
+	.hidden	__clzdi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__clzdi2
+	.type	__clzdi2, @function
+__clzdi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	move    $2, $4
+	sltiu   $8, $5, 1
+	movn    $2, $5, $5
+	sll     $8, $8, 5
+	srl     $3, $2, 31
+	plzcw   $2, $2
+	addiu   $2, $2, 1
+	movn    $2, $zero, $3
+	jr	$31
+	addu    $2, $2, $8
+	.set	macro
+	.set	reorder
+
+	.end	__clzdi2
+	.size	__clzdi2, .-__clzdi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/r5900/_clzsi2.S b/libgcc/config/mips/r5900/_clzsi2.S
--- a/libgcc/config/mips/r5900/_clzsi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_clzsi2.S	2026-06-03 18:52:43.209402557 +0300
@@ -0,0 +1,33 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__clzsi2
+	.hidden	__clzsi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__clzsi2
+	.type	__clzsi2, @function
+__clzsi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	srl     $3, $4, 31
+	plzcw   $2, $4
+	addiu   $2, $2, 1
+	jr      $31
+	movn    $2, $zero, $3
+
+	.set	macro
+	.set	reorder
+
+	.end	__clzsi2
+	.size	__clzsi2, .-__clzsi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/r5900/_ctzdi2.S b/libgcc/config/mips/r5900/_ctzdi2.S
--- a/libgcc/config/mips/r5900/_ctzdi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_ctzdi2.S	2026-06-03 18:52:54.231332914 +0300
@@ -0,0 +1,44 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__ctzdi2
+	.hidden	__ctzdi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__ctzdi2
+	.type	__ctzdi2, @function
+__ctzdi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	sltiu   $2, $4, 1
+	sltiu   $3, $5, 1
+	and     $8, $2, $3
+	move    $3, $5
+	sltiu   $9, $4, 1
+	movn    $3, $4, $4
+	sll     $9, $9, 5
+	subu    $2, $zero, $3
+	sll     $8, $8, 31
+	and     $2, $3, $2
+	sra     $8, $8, 31
+	addiu   $2, $2, -1
+	plzcw   $2, $2
+	xori    $2, $2, 31
+	addu    $2, $2, $9
+	jr	$31
+	or      $2, $2, $8	
+	.set	macro
+	.set	reorder
+
+	.end	__ctzdi2
+	.size	__ctzdi2, .-__ctzdi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/r5900/_ctzsi2.S b/libgcc/config/mips/r5900/_ctzsi2.S
--- a/libgcc/config/mips/r5900/_ctzsi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_ctzsi2.S	2026-06-03 18:53:03.345275330 +0300
@@ -0,0 +1,37 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__ctzsi2
+	.hidden	__ctzsi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__ctzsi2
+	.type	__ctzsi2, @function
+__ctzsi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	subu	$3, $0, $4
+	and	$3, $4, $3
+	addiu	$3, $3, -1
+	sltiu   $2, $4, 1
+	plzcw	$3, $3
+	sll     $2, $2, 31
+	xori	$3, $3, 31
+	sra     $2, $2, 31
+	jr	$31
+	or      $2, $2, $3
+	.set	macro
+	.set	reorder
+
+	.end	__ctzsi2
+	.size	__ctzsi2, .-__ctzsi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/r5900/_ffsdi2.S b/libgcc/config/mips/r5900/_ffsdi2.S
--- a/libgcc/config/mips/r5900/_ffsdi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_ffsdi2.S	2026-06-03 18:53:12.864215185 +0300
@@ -0,0 +1,40 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__ffsdi2
+	.hidden	__ffsdi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__ffsdi2
+	.type	__ffsdi2, @function
+__ffsdi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	move    $3, $5
+	sltiu   $8, $4, 1
+	movn    $3, $4, $4
+	sll     $8, $8, 5
+	subu    $2, $zero, $3
+	addiu   $8, $8, 32
+	and     $2, $3, $2
+	addiu   $2, $2, -1
+	plzcw   $2, $2
+	subu    $2, $8, $2
+	or      $8, $4, $5
+	jr	$31
+	movz    $2, $zero, $8
+	.set	macro
+	.set	reorder
+
+	.end	__ffsdi2
+	.size	__ffsdi2, .-__ffsdi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/r5900/_ffssi2.S b/libgcc/config/mips/r5900/_ffssi2.S
--- a/libgcc/config/mips/r5900/_ffssi2.S	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/r5900/_ffssi2.S	2026-06-03 18:53:20.694165711 +0300
@@ -0,0 +1,36 @@
+	.file	1 "libgcc2.c"
+	.section .mdebug.abi32
+	.previous
+	.nan	legacy
+	.module	softfloat
+	.module	oddspreg
+	.abicalls
+	.text
+	.align	2
+	.globl	__ffssi2
+	.hidden	__ffssi2
+	.set	nomips16
+	.set	nomicromips
+	.ent	__ffssi2
+	.type	__ffssi2, @function
+__ffssi2:
+	.frame	$sp,0,$31		# vars= 0, regs= 0/0, args= 0, gp= 0
+	.mask	0x00000000,0
+	.fmask	0x00000000,0
+	.set	noreorder
+	.set	nomacro
+	subu    $2, $zero, $4
+	and     $2, $4, $2
+	ori     $8, $zero, 32
+	addiu   $2, $2, -1
+	addu    $3, $4, $zero
+	plzcw   $2, $2
+	subu    $2, $8, $2
+	jr	$31
+	movz    $2, $zero, $3
+	.set	macro
+	.set	reorder
+
+	.end	__ffssi2
+	.size	__ffssi2, .-__ffssi2
+	.ident	"GCC: (GNU) 9.2.0"
diff -Naur a/libgcc/config/mips/t-r5900 b/libgcc/config/mips/t-r5900
--- a/libgcc/config/mips/t-r5900	1970-01-01 03:00:00.000000000 +0300
+++ b/libgcc/config/mips/t-r5900	2026-06-03 18:51:28.232873584 +0300
@@ -0,0 +1,15 @@
+LIB2FUNCS_EXCLUDE += _clrsbsi2 _clzsi2 _ctzsi2 _ffssi2
+
+LIB2ADD += $(srcdir)/config/mips/r5900/_clrsbsi2.S \
+           $(srcdir)/config/mips/r5900/_clzsi2.S \
+           $(srcdir)/config/mips/r5900/_ctzsi2.S \
+           $(srcdir)/config/mips/r5900/_ffssi2.S
+
+ifneq ($(filter -mabi=32,$(MULTILIB_CFLAGS) $(CFLAGS)),)
+LIB2FUNCS_EXCLUDE += _clrsbdi2 _clzdi2 _ctzdi2 _ffsdi2
+LIB2ADD += $(srcdir)/config/mips/r5900/_clrsbdi2.S \
+           $(srcdir)/config/mips/r5900/_clzdi2.S \
+           $(srcdir)/config/mips/r5900/_ctzdi2.S \
+           $(srcdir)/config/mips/r5900/_ffsdi2.S
+endif
+
diff -Naur a/libgcc/config.host b/libgcc/config.host
--- a/libgcc/config.host	2019-01-17 15:29:13.000000000 +0300
+++ b/libgcc/config.host	2026-06-19 07:11:09.629749398 +0300
@@ -917,6 +917,8 @@
 	  mips64r5900* | mipsr5900*)
 	    # The MIPS16 support code uses floating point
 	    # instructions that are not supported on r5900.
+	    # Optimize libgcc functions with special instructions as long as they comply with the ABI.
+	    tmake_file="${tmake_file} mips/t-r5900"
 	    ;;
 	  *)
 	    tmake_file="${tmake_file} mips/t-mips16 t-slibgcc-libgcc"
