diff -Naur gcc-9.2.0/libsanitizer/asan/asan_linux.cc gcc-9.2.0_fixed_include/libsanitizer/asan/asan_linux.cc
--- gcc-9.2.0/libsanitizer/asan/asan_linux.cc	2018-10-31 14:14:23.000000000 +0300
+++ gcc-9.2.0_fixed_include/libsanitizer/asan/asan_linux.cc	2020-03-12 18:28:36.053393035 +0300
@@ -31,6 +31,7 @@
 #include <dlfcn.h>
 #include <fcntl.h>
 #include <limits.h>
+#include <linux/limits.h>
 #include <pthread.h>
 #include <stdio.h>
 #include <unistd.h>
