diff -Naur flex-2.6.0_orig/tests/Makefile.am flex-2.6.0_edited/tests/Makefile.am
--- flex-2.6.0_orig/tests/Makefile.am	2015-11-17 19:16:43.000000000 +0300
+++ flex-2.6.0_edited/tests/Makefile.am	2018-12-28 08:58:37.540111228 +0300
@@ -332,7 +332,7 @@
 
 # specify how to process .l files in order to test the flex built by make all
 
-FLEX = $(top_builddir)/src/flex
+FLEX = /usr/bin/flex
 
 .l.c: $(FLEX)
 	$(FLEX) -o $@ $<
