diff -Naur libcap-2.24_orig/libcap/cap_file.c libcap-2.24_edited/libcap/cap_file.c
--- libcap-2.24_orig/libcap/cap_file.c	2013-12-24 21:35:13.000000000 +0400
+++ libcap-2.24_edited/libcap/cap_file.c	2018-12-28 11:48:14.825154238 +0300
@@ -10,7 +10,15 @@
 #include <sys/stat.h>
 #include <unistd.h>
 
+#ifndef XATTR_CAPS_SUFFIX
+#define XATTR_CAPS_SUFFIX "capability"
+#endif
+#ifndef XATTR_SECURITY_PREFIX
 #define XATTR_SECURITY_PREFIX "security."
+#endif
+#ifndef XATTR_NAME_CAPS
+#define XATTR_NAME_CAPS XATTR_SECURITY_PREFIX XATTR_CAPS_SUFFIX
+#endif
 
 #include "libcap.h"
 
diff -Naur libcap-2.24_orig/libcap/Makefile libcap-2.24_edited/libcap/Makefile
--- libcap-2.24_orig/libcap/Makefile	2014-01-06 04:55:03.000000000 +0400
+++ libcap-2.24_edited/libcap/Makefile	2018-12-28 11:45:18.354761648 +0300
@@ -16,8 +16,8 @@
 # no support).
 ifeq ($(LIBATTR),yes)
 FILES += cap_file
-LDFLAGS += -lattr
-DEPS = -lattr
+LDFLAGS +=
+DEPS =
 endif
 
 INCLS=libcap.h cap_names.h $(INCS)
@@ -43,7 +43,7 @@
 		$< >$@
 
 _makenames: _makenames.c cap_names.list.h
-	$(BUILD_CC) $(BUILD_CFLAGS) $< -o $@
+	gcc $(BUILD_CFLAGS) $< -o $@
 
 cap_names.h: _makenames
 	./_makenames > cap_names.h
diff -Naur libcap-2.24_orig/Make.Rules libcap-2.24_edited/Make.Rules
--- libcap-2.24_orig/Make.Rules	2014-01-06 05:16:21.000000000 +0400
+++ libcap-2.24_edited/Make.Rules	2018-12-28 11:44:37.507133716 +0300
@@ -5,6 +5,7 @@
 # common 'packaging' directoty
 
 FAKEROOT=$(DESTDIR)
+prefix=/usr/local/ps2/mipsr5900el-unknown-linux-gnu
 
 # Autoconf-style prefixes are activated when $(prefix) is defined.
 # Otherwise binaries and libraries are installed in /{lib,sbin}/,
@@ -48,12 +49,12 @@
 KERNEL_HEADERS := $(topdir)/libcap/include/uapi
 IPATH += -fPIC -I$(KERNEL_HEADERS) -I$(topdir)/libcap/include
 
-CC := gcc
+CC := mipsr5900el-unknown-linux-gnu-gcc
 CFLAGS := -O2 -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64
 BUILD_CC := $(CC)
 BUILD_CFLAGS := $(CFLAGS) $(IPATH)
-AR := ar
-RANLIB := ranlib
+AR := mipsr5900el-unknown-linux-gnu-ar
+RANLIB := mipsr5900el-unknown-linux-gnu-ranlib
 DEBUG = -g #-DDEBUG
 WARNINGS=-Wall -Wwrite-strings \
         -Wpointer-arith -Wcast-qual -Wcast-align \
