diff -Naur psmisc-22.20_orig/configure psmisc-22.20_edited/configure
--- psmisc-22.20_orig/configure	2012-09-19 16:29:26.000000000 +0400
+++ psmisc-22.20_edited/configure	2018-12-28 12:59:25.532253363 +0300
@@ -4118,13 +4118,13 @@
 if test "x$ac_cv_lib_tinfo_tgetent" = xyes; then :
   TERMCAP_LIB=-ltinfo
 else
-   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for tgetent in -lncurses" >&5
-$as_echo_n "checking for tgetent in -lncurses... " >&6; }
+   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for tgetent in -lncursesw" >&5
+$as_echo_n "checking for tgetent in -lncursesw... " >&6; }
 if ${ac_cv_lib_ncurses_tgetent+:} false; then :
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lncurses  $LIBS"
+LIBS="-lncursesw  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -4155,7 +4155,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_ncurses_tgetent" >&5
 $as_echo "$ac_cv_lib_ncurses_tgetent" >&6; }
 if test "x$ac_cv_lib_ncurses_tgetent" = xyes; then :
-  TERMCAP_LIB=-lncurses
+  TERMCAP_LIB=-lncursesw
 else
    { $as_echo "$as_me:${as_lineno-$LINENO}: checking for tgetent in -ltermcap" >&5
 $as_echo_n "checking for tgetent in -ltermcap... " >&6; }
@@ -7310,23 +7310,23 @@
 fi
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_func_malloc_0_nonnull" >&5
 $as_echo "$ac_cv_func_malloc_0_nonnull" >&6; }
-if test $ac_cv_func_malloc_0_nonnull = yes; then :
+#if test $ac_cv_func_malloc_0_nonnull = yes; then :
 
 $as_echo "#define HAVE_MALLOC 1" >>confdefs.h
 
-else
-  $as_echo "#define HAVE_MALLOC 0" >>confdefs.h
+#else
+#  $as_echo "#define HAVE_MALLOC 0" >>confdefs.h
 
-   case " $LIBOBJS " in
-  *" malloc.$ac_objext "* ) ;;
-  *) LIBOBJS="$LIBOBJS malloc.$ac_objext"
- ;;
-esac
+#   case " $LIBOBJS " in
+#  *" malloc.$ac_objext "* ) ;;
+#  *) LIBOBJS="$LIBOBJS malloc.$ac_objext"
+# ;;
+#esac
 
 
-$as_echo "#define malloc rpl_malloc" >>confdefs.h
+#$as_echo "#define malloc rpl_malloc" >>confdefs.h
 
-fi
+#fi
 
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for working memcmp" >&5
@@ -7439,23 +7439,23 @@
 fi
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_func_realloc_0_nonnull" >&5
 $as_echo "$ac_cv_func_realloc_0_nonnull" >&6; }
-if test $ac_cv_func_realloc_0_nonnull = yes; then :
+#if test $ac_cv_func_realloc_0_nonnull = yes; then :
 
 $as_echo "#define HAVE_REALLOC 1" >>confdefs.h
 
-else
-  $as_echo "#define HAVE_REALLOC 0" >>confdefs.h
+#else
+#  $as_echo "#define HAVE_REALLOC 0" >>confdefs.h
 
-   case " $LIBOBJS " in
-  *" realloc.$ac_objext "* ) ;;
-  *) LIBOBJS="$LIBOBJS realloc.$ac_objext"
- ;;
-esac
+#   case " $LIBOBJS " in
+#  *" realloc.$ac_objext "* ) ;;
+#  *) LIBOBJS="$LIBOBJS realloc.$ac_objext"
+# ;;
+#esac
 
 
-$as_echo "#define realloc rpl_realloc" >>confdefs.h
+#$as_echo "#define realloc rpl_realloc" >>confdefs.h
 
-fi
+#fi
 
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether stat accepts an empty string" >&5
diff -Naur psmisc-22.20_orig/src/fuser.c psmisc-22.20_edited/src/fuser.c
--- psmisc-22.20_orig/src/fuser.c	2012-09-19 14:54:03.000000000 +0400
+++ psmisc-22.20_edited/src/fuser.c	2018-12-28 12:59:25.532253363 +0300
@@ -37,6 +37,7 @@
 #include <sys/stat.h>
 #include <sys/socket.h>
 #include <sys/wait.h>
+#include <sys/sysmacros.h>
 #include <arpa/inet.h>
 #include <netinet/in.h>
 #include <pwd.h>
